Add computation and retrieval of batch feature statistics

zhilingc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. Ensure that your code follows our code conventions: https://github.com/gojek/feast/blob/master/docs/contributing.md#code-conventions
2. Run unit tests and ensure that they are passing: https://github.com/gojek/feast/blob/master/docs/contributing.md#running-unit-tests
3. If your change introduces any API changes, make sure to update the integration tests scripts here: https://github.com/gojek/feast/tree/master/tests/e2e
4. Make sure documentation is updated for your PR!
5. Make sure you have signed the CLA https://cla.developers.google.com/clas

-->

**What this PR does / why we need it**:
This PR adds support for retrieval of batch statistics over data ingested into a feast warehouse store, as proposed in M2 of the [Feature Validation RFC](https://docs.google.com/document/d/1TPmd7r4mniL9Y-V_glZaWNo5LMXLshEAUpYsohojZ-8/edit?pli=1#). 

Note that it deviates from the RFC in the following ways:
- Statistics are computed using SQL. This is because TFDV is unfortunately, only available in python, and Multi-SDK connectors for Beam is still a [work in progress](https://beam.apache.org/roadmap/connectors-multi-sdk/). Computing the statistics using SQL will be the compromise until either TFDV is available in Java, or cross-language execution is supported.
- Statistics can only be computed over a single feature-set at a time. This is mostly to reduce complexity in implementation. Since datasets are unable to span multiple feature sets, it makes sense to have this restriction in place. 

This is a bit of a chonky PR, and the code itself requires a bit of cleaning up, hence the WIP status, but refer to the attached [notebook](https://github.com/gojek/feast/blob/16867806944a9f8634c8503a118db6b1e9ed1ed9/examples/statistics/Historical%20Feature%20Statistics%20with%20Feast%2C%20TFDV%20and%20Facets.ipynb) for how this implementation looks like for a user of Feast.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information about release notes, see kubernetes' guide here:
http://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
- Adds GetFeatureStatistics to the CoreService
- Adds get_statistics method to the client in the python SDK
```

woop commented on core/src/main/java/feast/core/model/FieldId.java:
```diff
@@ -0,0 +1,68 @@
+@AllArgsConstructor
+@Getter
+@Setter
+public class FieldId implements Serializable {
```
Perhaps FieldRef or FeatureRef?

woop commented:
Will try to review it tomorrow morning @zhilingc

woop commented on core/src/main/java/feast/core/grpc/CoreServiceImpl.java:
```diff
@@ -113,6 +95,21 @@ public void listFeatureSets(
+      log.error("Exception has occurred in GetFeatureStatistics method: ", e);
+      responseObserver.onError(
+          Status.INTERNAL.withDescription(e.getMessage()).withCause(e).asRuntimeException());
+    }
```
Would it make sense to also return the message to the end user of other exceptions? It's obviously not a good practice long term, but for the time being it would make debugging easier.

woop commented on core/src/main/java/feast/core/model/Entity.java:
```diff
@@ -0,0 +1,127 @@
+
+  public Entity() {}
+
+  public Entity(FieldId fieldId) {
```
We should possibly make these static factory methods like `fromProto`. It would make the API a bit cleaner.

woop commented on sdk/python/feast/client.py:
```diff
@@ -851,6 +858,78 @@ def ingest(
+        return self._core_service_stub.GetFeatureStatistics(
+            request
+        ).dataset_feature_statistics_list
         return None
```
Should this `return None` be deleted?

zhilingc replied:
oops. might have been from rebasing

woop commented on sdk/python/feast/client.py:
```diff
@@ -908,6 +987,20 @@ def _build_feature_references(
     return features
 
 
+def _generate_dataset_id(feature_set: FeatureSet) -> str:
```
The term dataset still seems a bit off to me.

The `id` that we are returning is a `job_id` or an `ingestion_id` since it is unique to every invocatio of `ingest()`. If we refer to it as `dataset_id` then I think users would expect it to be associated with their dataset, which means they provide the value or at the very least the value doesnt change in subsequent runs.

zhilingc replied:
I was wary of letting users set this value since there is a chance they would assign the same value twice. And running the ingestion twice SHOULD have distinct ids because users probably don't want to treat the data as belonging to the same set... I can add the option to override the dataset id, but it really opens up the possibility of users breaking the system.

I don't think it should be a `job_id`, which is already an extremely overloaded term, and retrieval of data by `ingestion_id` doesn't seem as intuitive as something like a `dataset_id`

woop replied:
The reason I brought this up is because the current implementation does not fit the typical 
SOP for our users where they build idempotent data system. They expect to be able to create a dataset and ingest it based on a name, and have the data system upsert. So basically they just run ingestion until it succeeds and then they are done. They don't care, and often don't know, that we are running Kafka.

And our users definitely don't want to maintain a list of UUIDs. They probably just want to query their datasets by the names they are using like "2020417-mydataset`

Regardless, I think your current approach is fine for now because nothing prevents us from eventually allowing them to set this Id if we allow idempotency. I would probably reuse this same `dataset_id` though.

zhilingc replied:
ok. I've added the option for people to supply their own dataset id, and fixed the sql so that it only retrieves the latest value for each dataset/entity/event timestamp. This should ensure idempotency.

woop commented on storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/statistics/FeatureSetStatisticsQueryInfo.java:
```diff
@@ -0,0 +1,101 @@
+  private String date = "";
+  private final List<FieldStatisticsQueryInfo> features;
+
+  public FeatureSetStatisticsQueryInfo(
```
Is this constructor being used?

zhilingc replied:
Nope. Removed.

woop commented on storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/statistics/FieldStatisticsQueryInfo.java:
```diff
@@ -0,0 +1,94 @@
+/**
+ * Value class for Features containing information necessary to template stats-retrieving queries.
+ */
+public class FieldStatisticsQueryInfo {
```
Just my own curiousity. How are we deciding when to use AutoValue and when to not use AutoValue?

In this case we have a value class which seems like it fits the AutoValue use case well, but we opt not to use it. Then for complex classes like `BigQueryStatisticsRetriever` we do opt for AutoValue.

zhilingc replied:
I personally use autovalue when there are classes that require more than 2 elements to construct, but pebble is unfortunately incompatible with autovalue classes, so in this case I can't.

woop commented on storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/statistics/FieldStatisticsQueryInfo.java:
```diff
@@ -0,0 +1,94 @@
+      case STRING_LIST:
+        return new FieldStatisticsQueryInfo(entitySpec.getName(), "LIST");
+      default:
+        throw new IllegalArgumentException("Invalid entity type provided");
```
Can we return the `valueType` as a string?

zhilingc replied:
Sorry, what do you mean?

woop replied:
I mean can we add the case value into the exception so that it's easier to debug.

woop commented on tests/e2e/feature-validation.py:
```diff
@@ -0,0 +1,99 @@
+    )
+
+    expected_stats = statistics_pb2.DatasetFeatureStatisticsList()
+    with open("statistics/expected_output_basic_dataset.json", "r") as fo:
```
I think it would be less fragile if we generated the statistics from the dataframe using TFDV instead of reading it from a file. That would make it a lot easier to ensure that we have and can maintain compatibility.

zhilingc replied:
We don't support the full set of TFDV stats, so generating it on the fly wouldn't allow the test to pass.

woop replied:
Possibly to filter out the ones we don't need?

woop commented on tests/e2e/feature-validation.py:
```diff
@@ -0,0 +1,99 @@
+    stats = client.get_statistics(
+        feature_refs=["strings", "ints", "floats"],
+        store="bigquery",
+        start_date=dataset_basic["date"],
```
This just tests a single day right?

zhilingc replied:
yes

woop commented:
I made a first run through.

The only big thing that stands out is the duplication between entity level statistics and feature level statistics. I'll think about how we can simplify this implementation because as written it is a lot of scary to merge this in.

Other than that I am finding it a bit slow going to review the PR because some of the methods require comments. I'd rather not ding the methods and ask for JavaDocs or comments, so would you mind just giving the PR a scan through and adding them? It would be especially useful for `StatsService.java` as well as the model fields.

woop commented:
I'll talk about the data model first because that is the most important part to get right now.

**Assumptions**
- Entities are primarily used as keys
- Features are primarily used as values in the context of models. Our users care more about statistics on features than on entities.
- Features need references (like feature references).
- Entities need references, but these references are different from features.
- Entities will also be treated differently in the future (See #405). Constraints on features and constraints on entities are different. Entities will be reused and features must be unique by name.
- Nothing prevents a user from ingesting entity data as a feature. 

**Thoughts**
- Although there is a use case for having entity level statistics, I don't think it warrants the amount of code we'd need to maintain here.
- There are pre-existing problems with the data model that should be solved now before 0.5 is released. Normally that would happen in a separate PR, but we can also reuse this PR if that is easier for you. 
- One of the problems that I see is the fact that `Field` and its inheritence adds complexity without any real value.
- The second is that there is a coupling between the references of features and entities that I don't think should exist.
- Lastly, and this is definitely out of scope for your work, is that the source table is hardcoded to Kafka (thanks for the reminder @mrzzy . Will add a separate issue for that.)

**Recommendation**
- Remove `Field` and its inheritance
- Keep `FieldId` but rename it to `FeatureReference`
- Have `Feature` reference `FeatureReference` as its primary key
- Create an `EntityReference` (for the time being this can be the same as `FeatureReference`)
- Keep `FeatureStatistics` and have it reference `FeatureReference`
- Remove `EntityStatistics` completely as well as statistics methods on Entity.

cc @ches @Yanson

zhilingc commented:
@woop 

> Features are primarily used as values in the context of models. Our users care more about statistics on features than on entities.
> Although there is a use case for having entity level statistics, I don't think it warrants the amount of code we'd need to maintain here.

I think that’s a very strong assumption to make. 

I’ve been back and forth about entity statistics the past few weeks; my first iteration only involved feature statistics, but i ended up adding support for entity statistics later because i think it is something that is truly valuable to users, for checking id distributions and bounds, etc. I don’t think it’s 

To your point:

> Nothing prevents a user from ingesting entity data as a feature.

This doesn’t make sense when you consider the user experience. Do they double-up on entity columns so that they can be ingested as features JUST to retrieve statistics? Do they ingest it as a separate dataset? Why do users have to perform hacks to transform a logical unit (the dataset) into something else just to do something that is expectedly a basic requirement?

> One of the problems that I see is the fact that Field and its inheritance adds complexity without any real value.

I’m not sure what sort of complexity you mean here. I think if we do have statistics for entities it’s a lot easier to maintain compatibility with TFDV fields with inheritance, which is the main reason why the class was introduced in the first place. Additional fields that differ between features and entities can be defined in their concrete classes.

> The second is that there is a coupling between the references of features and entities that I don't think should exist.

Do you mean the fact that they use the same object? I guess it’s fine to separate them for the sake of forward compatibility…

> Have Feature reference FeatureReference as its primary key
Do you mean have FeatureReference as a distinct entity and have feature sets, features and statistics reference that entity? I’m not sure if that’s the best idea.

> Keep FeatureStatistics and have it reference FeatureReference
FeatureReference is part of the definition of Feature and FeatureStatistics cannot form a relation directly to it. To be exact, with the current implementation, the table definition for Features looks like:

```
-- Table Definition ----------------------------------------------

CREATE TABLE feature (
    feature_set character varying(255),
    name character varying(255),
    project character varying(255),
    version integer,
    bool_domain bytea,
    domain character varying(255),
    float_domain bytea,
    group_presence bytea,
    image_domain bytea,
    int_domain bytea,
    mid_domain bytea,
    natural_language_domain bytea,
    presence bytea,
    shape bytea,
    string_domain bytea,
    struct_domain bytea,
    time_domain bytea,
    time_of_day_domain bytea,
    type character varying(255),
    url_domain bytea,
    value_count bytea,
    CONSTRAINT feature_pkey PRIMARY KEY (feature_set, name, project, version)
);
```

And FeatureStatistics has the following constraint:
```
    CONSTRAINT fksrjnxx9em2xg2cqgd6ifm4lll FOREIGN KEY (feature_set, name, project, version) REFERENCES feature(feature_set, name, project, version)
```

zhilingc commented:
Ok, I polled a few data scientists and they acknowledged that there is no need to have entity statistics except for occasional once-off checks, which can be done directly to the db anyway. I'll remove the entity statistics - but does that mean i remove the entity field schemas introduced in #438 as well?

woop commented:
> I’ve been back and forth about entity statistics the past few weeks; my first iteration only involved feature statistics, but i ended up adding support for entity statistics later because i think it is something that is truly valuable to users, for checking id distributions and bounds, etc. I don’t think it’s

I think your sentence cut off there.

I agree that entity statistics are very useful, don't get me wrong, which is why we wanted them in the first place.

What doesn't line up for me is this implementation. As it stands, Feast does not support entity statistics. If I am forced to choose between having only feature statistics and merging in all of this code then I will definitely chose the former.

However, the idea case is that we can **cleanly** support entity statistics.

> This doesn’t make sense when you consider the user experience. Do they double-up on entity columns so that they can be ingested as features JUST to retrieve statistics? Do they ingest it as a separate dataset? Why do users have to perform hacks to transform a logical unit (the dataset) into something else just to do something that is expectedly a basic requirement?

My point here is that we can add feature statistics safely and support it properly without immediately supporting entity statistics. If users really want a work-around to get an idea of entity statistics then they can ingest them as features, which means adding additional columns. That would not be something we would officially support, and they don't lose anything over what they have today.

Duplicating this code for both entities and features is after all tech debt. Another Feast engineer will have to clean this up later. I would rather solve this properly or scope it down.

> I’m not sure what sort of complexity you mean here. I think if we do have statistics for entities it’s a lot easier to maintain compatibility with TFDV fields with inheritance, which is the main reason why the class was introduced in the first place. Additional fields that differ between features and entities can be defined in their concrete classes.

I stand corrected.

> Do you mean the fact that they use the same object? I guess it’s fine to separate them for the sake of forward compatibility…

Yes. Their constraints are distinct.

> Do you mean have FeatureReference as a distinct entity and have feature sets, features and statistics reference that entity? I’m not sure if that’s the best idea.

Exactly how it is right now for FieldId, but it's just FeatureReference. Entity doesn't reference FeatureReference. It doesn't have to be a distinct entity, it can be an embeddedId.

> FeatureReference is part of the definition of Feature and FeatureStatistics cannot form a relation directly to it. To be exact, with the current implementation, the table definition for Features looks like:

The table definition you are showing me is exactly what I want. It's a foreign key based on multiple foreign columns on the feature reference (name, feature_set, project, version). I guess it just comes down to a naming difference (`FieldId` -> `FeatureReference`). 

--------------

It's very clear that our existing data model is wrong because it forces your statistics design to be split between features and entities. Statistics are ultimately about data, and the only common concepts in our data layer is the `Field`.

![image](https://user-images.githubusercontent.com/6728866/79743727-ac935c80-8337-11ea-9049-41e622476eca.png)

So our options are

**1. Map statistics to Fields instead of Entities and Features.**
This does make sense at the level of feature statistics. Features map to data through fields. Entities map to data through fields. Statistics occur on the data/field.

The problem with our current `Field` is that the parent child relationship of fields to features and entities doesn't work. `Field` as a means of sharing common Java fields through inheritance makes sense, but `Fields` shouldn't be the parents of `Entities` and `Features` because `Entities` don't share an identity with a `Field`. Each `Field` in a `FeatureRow` is unique, and that `Field`, even if its an `Entity`, has a different identity for each `Feature Set`. 

One way to make this work is invert the relationship between Feature/Entity and Field to one that is consistent with [FeatureRow.proto](https://github.com/gojek/feast/blob/master/protos/feast/types/FeatureRow.proto)

* `Fields` have an identity (`FieldRef`) and a table.
* `Feature set` has a single collection of `Fields` (instead of `Features` and `Entities`). This creates a many to many mapping between `Feature Set` and `Field`
* A `Field` can then be configured to be one of two types, `Entity` or a `Feature`. 
* `Fields` that have a `Feature` type don't have an external identity. They share an identity with the `Field`, `FieldRef`. There may or may not be a `Feature` table that references this `Field` table. Not sure if it's necessary.
* `Fields` that have an `Entity` type reference an external `Entity`. Meaning an `Entity` has a table. These entities can occur in multiple `Feature Sets`. This might be out of scope for your PR though.
* Statistics is successfully de-duplicated and only reference a `Field`

**2. Replace Field with Feature** 
The thing with Proposal (1) is that the identity of `Field` will probably be `FieldId`/`FieldRef` which is 100% identical to `FeatureRef`. That raises the question of what differentiates these two (Field/Feature). Do we really need `FieldId` and `FeatureRef`? 

If we want to treat all `Fields` as `Features` (generating statistics or feeding them into models), then we could simplify the data model by replacing `Field` with `Feature` in proposal 1. That removes the duplication between `FieldId` and `FeatureRef`. 

**3. Just remove Field** 
The final proposal is the one I gave you previously. We remove `Field` and keep `Entities` and `Features` separate and allow statistics only on `Features`. This is the smallest cut that we can make right now, since the data model doesn't dramatically change.

Ok, so there you have three proposals and my logic behind it. Please come up with a solution

zhilingc commented:
The issue with (1) is that features were split to be distinct from entities because while they're currently mostly congruent to each other, that won't be the case in the future. 
I'll do as you suggested in (3), but does that mean i remove the entity field schemas introduced in #438 as well?

woop commented:
> The issue with (1) is that features were split to be distinct from entities because while they're currently mostly congruent to each other, that won't be the case in the future.

Ok, how does (1) prevent that? All of the options keep entities and features disconnected.

> I'll do as you suggested in (3), but does that mean i remove the entity field schemas introduced in #438 as well?

You would need to remove entity schemas yes. I am fine with going for (3), but I would still like to hear your thoughts on what I wrote above.

woop commented on core/src/main/java/feast/core/model/Entity.java:
```diff
@@ -19,94 +19,35 @@
   }
 
-  public static Entity withId(FieldId fieldId) {
+  public static Entity withRef(EntityReference entityId) {
```
`entityRef` perhaps?

woop commented on core/src/main/java/feast/core/model/Feature.java:
```diff
@@ -32,19 +32,49 @@
   private Feature(String name, ValueType.Enum type) {
-    this.setId(new FieldId());
-    this.setName(name);
+    this.setReference(new FeatureReference());
```
Is there a reason to have a no args constructor for `FeatureReference`? If there isn't then I think it might be safer to require all the fields/args that make up the reference.

zhilingc replied:
Using a constructor with just name now. The previous model refactor you did injects the project and feature set name independently of the initialisation of the Feature/Entity class. I don't think it's within scope to change that right now.

woop commented on core/src/main/java/feast/core/model/Feature.java:
```diff
@@ -32,19 +32,49 @@
   }
 
-  public static Feature withId(FieldId fieldId) {
+  public static Feature withReference(FeatureReference featureId) {
```
is `with` used in practice? How about `fromRef`?

zhilingc replied:
`fromRef` doesn't convey the meaning though. It's called a `reference` but its more of a property of the class

woop commented on core/src/main/java/feast/core/service/StatsService.java:
```diff
@@ -0,0 +1,587 @@
+      long timestamp = request.getStartDate().getSeconds();
+      while (timestamp < request.getEndDate().getSeconds()) {
+        List<FeatureNameStatistics> featureNameStatistics =
+            getFeatureNameStatisticsByDate(
```
Would this results in multiple calls to the Feast db and multiple calls to BigQuery? If so, can we use the full time range?

zhilingc replied:
No, if the full time range is used, the values cannot be cached.

woop replied:
Correct me if I am wrong, but wouldn't the only requirement that the query we send to BigQuery returns multiple results? 

Meaning we check our cache (in one query), we ask BQ for the days that we don't have (in one query), we cache them, and then we return the full set.

woop commented on core/src/main/java/feast/core/service/StatsService.java:
```diff
@@ -0,0 +1,587 @@
+              "Invalid store %s specified. Batch statistics are only supported for BigQuery stores",
+              store.getType()));
+    }
+    return BigQueryStatisticsRetriever.newBuilder()
```
Can we please use a static factory method to make this consistent with the other retrievers?

`fromConfig(Map<String, String> config)`?

zhilingc replied:
done

woop commented on core/src/main/java/feast/core/model/FeatureStatistics.java:
```diff
@@ -0,0 +1,269 @@
+  })
+  private Feature feature;
+
+  // Only one of these fields should be populated.
```
Previously when we discussed this we said we would store statistics with a key of `day` + `dataset`, which would allow us to roll it up to any time range and to any dataset.

Is there a reason why both won't be set here? Are we still doing that?

zhilingc replied:
a large portion of statistics are lost in aggregation, so we want to avoid aggregating within datasets, which should be the logical unit for retrieval here anyway.

woop replied:
Have we documented the list of statistics that are lost? How are we dealing with that in the case of a time range query?

woop commented on core/src/main/java/feast/core/model/FeatureStatistics.java:
```diff
@@ -0,0 +1,269 @@
+
+  // Instantiates a Statistics object from a tensorflow metadata FeatureNameStatistics object and a
+  // dataset ID.
+  public static FeatureStatistics fromProto(
```
I think we should perhaps change the method names in this class. `fromProto` and `toProto` are normally reserved for `domain model` conversion to and from our proto representation. We are passing non-proto methods here.

zhilingc replied:
`FeatureNameStatistics` is a proto.

woop commented on core/src/main/java/feast/core/model/FeatureStatistics.java:
```diff
@@ -0,0 +1,269 @@
+    return featureNameStatisticsBuilder.build();
+  }
+
+  private static FeatureStatistics fromProto(FeatureNameStatistics featureNameStatistics)
```
It's confusing that this overlaps with `fromProtos` above. It could have been a non-static method like `updateFromFeatureNameStatistics`

woop commented on core/src/main/java/feast/core/model/FeatureStatistics.java:
```diff
@@ -0,0 +1,269 @@
+    return featureStatistics;
+  }
+
+  public FeatureNameStatistics toProto() throws InvalidProtocolBufferException {
```
Same as the other comments. I think it should be renamed to indicate that this is exporting a TensorFlow `FeatureNameStatistics` object.

woop commented on core/src/main/java/feast/core/service/StatsService.java:
```diff
@@ -0,0 +1,582 @@
+
+  private FeatureSetSpec getFeatureSetSpec(String featureSetId)
+      throws InvalidProtocolBufferException, IllegalArgumentException, RetrievalException {
+    String[] split = featureSetId.split("/");
```
Is it possible for us to dedupe this feature ref conversion? Can we move it to `FeatureReference.java` so that other parts of the code base can just reuse that `toString`, `fromString` maybe?

zhilingc replied:
It's for the FeatureSet, not Feature. And doing a code search finds only a single usage of this in feast core, so I don't think refactoring it out is too important right now

woop commented:
Questions
* Do we have or can we add a multiple day (time) statistics aggregation test?
* Do we have a definitive list of statistics that we provide users?
* Do we have a list of statistics that they lose when they do aggregation?
* Are we testing that the results contain the above statistics?
* One of the biggest pains with our current bigquery storage implementation is that failures don't bubble up to the user. Can we also test that our exceptions bubble up? Especially if there is a problem with BigQuery configuration.
* Is there a test for the force refresh?
* Is there a test for the user providing no dataset/time range or providing both?

zhilingc commented:
> Do we have or can we add a multiple day (time) statistics aggregation test?

Added to the end to end tests.

> Do we have a definitive list of statistics that we provide users?
> Do we have a list of statistics that they lose when they do aggregation?

All in the gitbook documentation already. See the table at the bottom of the page.

> Are we testing that the results contain the above statistics?

Yes, in the end to end test.

> One of the biggest pains with our current bigquery storage implementation is that failures don't bubble up to the user. Can we also test that our exceptions bubble up? Especially if there is a problem with BigQuery configuration.

The grpc method catches all exceptions, so it will be bubbled up. I've tested this personally.

> Is there a test for the force refresh?

No, i've only tested this locally. I can add the unit test, but it will take time.

> Is there a test for the user providing no dataset/time range or providing both?

Added.

woop commented on protos/feast/core/CoreService.proto:
```diff
@@ -274,3 +281,42 @@ message StopIngestionJobRequest {
+
+    // Optional filter which filters returned statistics by selected features. These
+    // features must be present in the data that is being processed.
+    repeated string feature_ids = 2;
```
Feature references?

zhilingc replied:
feature names actually

woop replied:
Alright, so we change it to feature_names then?

woop replied:
I guess it doesn't matter much since we can always change a field name

zhilingc replied:
I think since we're providing `feature_set_id` `features` should logically refer to feature names within that feature set.

zhilingc commented:
/test test-end-to-end

zhilingc commented:
/ok-to-test

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end

zhilingc commented:
/test test-end-to-end

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

zhilingc commented:
/test test-end-to-end-batch

ches commented:
Would it be feasible to break this pull request up into a few somehow? At a first glance there's a lot that I think is generally beneficial independent of the feature addition of batch statistics, like entity/domain model additions and updates like `Entity`, feature and entity references, etc.

I have little doubt that the batch statistics feature addition in itself will necessarily be a large PR no matter how we slice it, but I see the 5,000 line diffstat here and my spirits get deflated that I will need to reserve the majority of a weekend to it sometime in the next few weeks, by which point it will be merged (which is totally fine, just that it feels like a drag to bring people back to it later if there are questions & comments).

ches commented:
Some products alternate technical and feature releases, starting to feel Feast 0.5 could be split on its two roadmap categories that way… speaking from the perspective of someone who will upgrade a deployment at some point (i.e. merging down a branch), [the scope is getting pretty huge](https://github.com/gojek/feast/compare/v0.4.7...master) (and that's not including this, #554, or #533 yet). It's great work, it's just creating some anxiety for me about the effort and risks of upgrading.

Anyway, I'm off topic for this PR, I'll take this elsewhere…

zhilingc commented:
@ches That's fair. I'll do what I can.

woop commented:
> Some products alternate technical and feature releases, starting to feel Feast 0.5 could be split on its two roadmap categories that way

Yea the scope of 0.5 has increased a lot, which is partly good and bad. By splitting do you mean based on the existing master branch versus the additional functionality we are introducing with open PRs, or would you move out some functionality also merged into master?

My intuition is to put our efforts towards making 0.5 as stable and well tested as possible, so we would probably deploy it side by side and migrate folks over slowly. I would want to include all breaking changes for that matter. I think its possible to have 0.5 and 0.6 be technical and feature releases, but we probably will roll up to 0.6 quite quickly, which means we won't have much time to maintain the 0.5 branch.
